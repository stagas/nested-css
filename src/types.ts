// TODO: docs are not rendering correctly for this type
// https://github.com/TypeStrong/typedoc/issues/1519
/**
 * Nested CSS rules.
 */
export type NestedCSSDeclaration = Partial<CSSStyleDeclaration> & {
  [name: string]: string | NestedCSSDeclaration
}
