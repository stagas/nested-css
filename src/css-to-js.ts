import { createTokenizer } from 'tokenizer-next'
import type { NestedCSSDeclaration } from './types'

const extraWhitespace = /\s+/g

const tokenizer = createTokenizer(
  /(?<open>\s*\{\s*)/,
  /(?<close>\s*\}\s*)/,
  /(?<comment>\/\*[^]*?\*\/)/,
  /\s*(?<rule>[^{;}/'"]+)\s+?(?={)/,
  /\s*(?<string>'.*?'|".*?")(?=;)/,
  /\s*(?<prop>[^:;/]+)(?=:)/,
  /\s*(?<value>[^:;]+)(?=;)/,
)

/**
 * Convert a CSS string to a {@link NestedCSSDeclaration}.
 *
 * @param input
 */
export function cssToJs(input: string) {
  input = input.replace(extraWhitespace, ' ')

  const nextToken = tokenizer(input)
  const filter = (ignored: string[]) => (() => {
    let token
    while (token = nextToken())
      if (!ignored.includes(token.group)) break
    return token
  })
  const next = filter(['comment'])

  const parse = (style: NestedCSSDeclaration = {}) => {
    let token

    while ((token = next())) {
      const { value, group } = token

      switch (group) {
        case 'rule':
          next() // open {
          // @ts-ignore can't handle the ??= for some reason
          parse(style[value.trim()] ??= {})
          break
        case 'prop':
          style[value.trim()] = (next()?.value as string)?.trim()
          break
        case 'close':
          return style
      }
    }

    return style
  }

  return parse()
}
