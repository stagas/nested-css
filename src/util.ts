/**
 * Joins parts with values
 *
 * @param parts
 * @param values
 */
export function joinPartsWithValues(parts: TemplateStringsArray, values: unknown[]) {
  let str = ''
  for (let i = 0; i < parts.length; i++) {
    str += parts[i]
    // if (i < values.length) str += values[i]
    str += values[i]
  }
  return str
}

/**
 * Convert string to kebab-case.
 *
 * @param s PascalCase or camelCase string
 */
export function kebabCase(s: string) {
  return s.replace(/[A-Z]/g, (m: string, i: number) => (i ? '-' : '') + m).toLowerCase()
}
