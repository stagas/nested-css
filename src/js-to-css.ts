// original: https://gist.github.com/mindplay-dk/47bc597dcbc28bea009b02c5a0c9010e

import type { NestedCSSDeclaration } from './types'
import { kebabCase } from './util'

const mediaRule = (rule: string, media: string) => (media ? `${media}{${rule}}` : rule)

const createRule = (target: string, prop: string, value: string) => `${target}{${kebabCase(prop)}:${value}}`

// const isClassOrRoot = /^\s*[.:]/g

const remap = (target: string, map?: Map<string, string>) =>
  // isClassOrRoot.test(target)
  //   ? target
  //   :
  [...target.matchAll(/[\w-]+|./g)]
    .flat()
    .map(x => map?.get(x) ?? x)
    .join('')

/**
 * Compile a JS nested rules {@link NestedCSSDeclaration} to a CSS string.
 *
 * Examples:
 * ```ts
 * jsToCss({ '.foo': { color: 'blue' } })
 * // => .foo{color:blue}
 *
 * // custom root
 * jsToCss({ color: 'red' }, '.my-button')
 * // => .my-button{color:red}
 *
 * // with substitution
 * jsToCss({ '.foo': { color: 'blue' } }, null, new Map([['foo', 'bar']]))
 * // => .bar{color:blue}
 * ```
 *
 * @param rules Rules object
 * @param rootSelector Top level rules will use this selector
 * @param aliasMap Alias identifiers (i.e for `.foo` to become `.bar`
 *  you will need a `foo=bar` entry)
 * @returns The compiled CSS string
 */
export function jsToCss(
  rules: NestedCSSDeclaration,
  rootSelector?: string | null | undefined,
  aliasMap?: Map<string, string>,
) {
  rootSelector ??= ''

  let css = ''

  const parse = (rules: NestedCSSDeclaration, child = '', media = '') => {
    for (const key in rules) {
      const value = rules[key]

      if (typeof value === 'object') {
        const isMedia = /^@/.test(key) ? key : null

        let target = child

        if (!isMedia) {
          target = key
            .split(',')
            .map(k => k.trim())
            .map(k =>
              child
                .split(',')
                .map(x => (x + (/^&/.test(k) ? k.slice(1) : ' ' + k)).trim())
                .join(',')
            )
            .map(k => k.trim())
            .join(',')
        }

        parse(value, target, isMedia || media)
      } else {
        const rule = mediaRule(
          createRule(child ? remap(child, aliasMap) : rootSelector!, key, value),
          media,
        )
        css += '\n' + rule.trim()
      }
    }
  }

  parse({
    [rootSelector]: rules,
  } as NestedCSSDeclaration)

  return css.trim()
}
