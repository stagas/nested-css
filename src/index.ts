export type { NestedCSSDeclaration } from './types'

import { cssToJs } from './css-to-js'
import { jsToCss } from './js-to-css'
import { joinPartsWithValues } from './util'

/**
 * Compile to CSS passing parameters to {@link jsToCss}.
 */
export type NestedCSSCompiler = ((
  rootSelector?: string,
  aliasMap?: Map<string, string>
) => string) & { valueOf: NestedCSSCompiler }

/**
 * Factory a {@link NestedCSSCompiler} for the given string.
 *
 * @param parts
 * @param values
 */
export function css(parts: TemplateStringsArray, ...values: unknown[]): NestedCSSCompiler {
  const nestedCssString = joinPartsWithValues(parts, values)
  const nestedCssObject = cssToJs(nestedCssString)

  function compileCss(rootSelector = ':host', aliasMap?: Map<string, string>) {
    return jsToCss(nestedCssObject, rootSelector, aliasMap)
  }

  compileCss.valueOf = compileCss

  return compileCss
}

export { cssToJs, jsToCss }
