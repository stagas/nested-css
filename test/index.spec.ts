import { css } from '../src'

describe('css', () => {
  it('should compile a nested css string to css', () => {
    const style = css`
      .some {
        .nested {
          color: blue;
        }
        &:hover {
          color: red;
        }
      }

      #someid {
        background: yellow;
      }
    `

    expect(style).toBeInstanceOf(Function)

    expect(style('')).toEqual(`.some .nested{color:blue}
.some:hover{color:red}
#someid{background:yellow}`)
  })

  it('valueOf() should be the compile function', () => {
    const style = css`
      .some {
        .nested {
          color: blue;
        }
        &:hover {
          color: red;
        }
      }

      #someid {
        background: yellow;
      }
    `

    expect(style).toBeInstanceOf(Function)
    expect(style.valueOf).toBe(style)

    expect(style.valueOf('')).toEqual(`.some .nested{color:blue}
.some:hover{color:red}
#someid{background:yellow}`)
  })

  it('replace given host', () => {
    const style = css`
      color: red;
      & {
        color: blue;
        .another {
          color: yellow;
        }
      }
    `
    expect(style('foo')).toEqual(`foo{color:red}
foo{color:blue}
foo .another{color:yellow}`)
  })

  it('should compile complex nested css string', () => {
    const style = css`
      :host {
        color: magenta;
      }

      :root {
        color: blue;
        background: green;
        /* comment */
      }
      /* comment */
      .foo,
      .bar {
        color: red;

        &:hover,
        &:focus {
          color: yellow;

          span {
            display: none; /* comment */
          }
        }
      }
      deep .selector:hover foo::after {
        margin: 0 auto;

        .deep {
          .nested {
            &:focus {
              visibility: hidden;
            }
          }
        }

        content: 'continues';
      }
    `

    expect(style('')).toEqual(`\
:host{color:magenta}
:root{color:blue}
:root{background:green}
.foo,.bar{color:red}
.foo:hover,.bar:hover,.foo:focus,.bar:focus{color:yellow}
.foo:hover span,.bar:hover span,.foo:focus span,.bar:focus span{display:none}
deep .selector:hover foo::after{margin:0 auto}
deep .selector:hover foo::after .deep .nested:focus{visibility:hidden}
deep .selector:hover foo::after{content:'continues'}`)
  })

  it('keyframes', () => {
    const style = css`
      @keyframes blink {
        0% {
          filter: brightness(1);
        }
        50% {
          filter: brightness(1.5);
        }
        100% {
          filter: brightness(2);
        }
      }
    `
    expect(style(':host')).toMatchSnapshot()
  })
})
