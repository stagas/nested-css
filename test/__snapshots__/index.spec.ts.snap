// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css keyframes 1`] = `
"@keyframes blink{0%{filter:brightness(1)}}
@keyframes blink{50%{filter:brightness(1.5)}}
@keyframes blink{100%{filter:brightness(2)}}"
`;
